queries/atm.py

```diff
@@ -0,0 +1,571 @@
+
+def QAtmClosest(node: Node, params: QueryStateDict, result: Result) -> None:
+    result.qkey = "ClosestAtm"
+
+
+def QAtmClosestDeposit(node: Node, params: QueryStateDict, result: Result) -> None:
+    result.qkey = "ClosestAtmDeposit"
+
+
+def QAtmClosestForeignExchange(
+    node: Node, params: QueryStateDict, result: Result
+) -> None:
+    result.qkey = "ClosestAtmForeignExchange"
+
+
+def QAtmClosestCoinmachine(node: Node, params: QueryStateDict, result: Result) -> None:
+    result.qkey = "ClosestAtmCoinmachine"
+
+
+def QAtmFurtherInfo(node: Node, params: QueryStateDict, result: Result) -> None:
+    """Reference to previous ATM query"""
+    q = result.state.get("query")
+    ctx = None if q is None else q.fetch_context()
+    if ctx is None or "result" not in ctx:
+        # There is a reference to a previous result
+        # which is not available: flag an error
+        result.error_context_reference = True
+    else:
+        result.context_reference = True
+        result.last_atm = ctx["result"]
+
+
+def QAtmFurtherInfoDeposit(node: Node, params: QueryStateDict, result: Result) -> None:
+    result.qkey = "AtmFurtherInfoDeposit"
+
+
+def QAtmFurtherInfoWithdrawalLimit(
+    node: Node, params: QueryStateDict, result: Result
+) -> None:
+    result.qkey = "AtmFurtherInfoWithdrawalLimit"
+
+
+def QAtmFurtherInfoOpeningHours(
+    node: Node, params: QueryStateDict, result: Result
+) -> None:
+    result.qkey = "AtmFurtherInfoOpeningHours"
+
+
+def QAtmFurtherInfoForeignExchange(
+    node: Node, params: QueryStateDict, result: Result
+) -> None:
+    result.qkey = "AtmFurtherInfoForeignExchange"
+
+
+def QAtmFurtherInfoCoinmachine(
+    node: Node, params: QueryStateDict, result: Result
+) -> None:
+    result.qkey = "AtmFurtherInfoCoinmachine"
+
+
+def _read_isb_location_data_from_file() -> List[Dict[str, Any]]:
+    """Read isb locations JSON data from file"""
+    try:
+        script_dir = os.path.dirname(__file__)
+        rel_path = _PATH_TO_ISB_JSON
+        abs_file_path = os.path.join(script_dir, rel_path)
+        with open(abs_file_path, "r", encoding="utf-8") as f:
+            return json.load(f)
+    except Exception as e:
+        logging.error("Error reading isb locations json data: %s", e)
+        return []
+
+
+def _get_atm_data() -> Optional[List[Dict[str, Any]]]:
+    """Get ATM data"""
+    atm_data: List[Dict[str, Any]] = _read_isb_location_data_from_file()
+    return atm_data
+
+
+def _atms_with_distance(loc: Optional[LatLonTuple]) -> Optional[List[Dict[str, Any]]]:
+    """Return list of atms w. added distance data."""
+    atm_data: Optional[List[Dict[str, Any]]] = _get_atm_data()
+    if not atm_data:
+        return None
+
+    filtered_atm_data: List[Dict[str, Any]] = []
+    for s in atm_data:
+        """Filter out all non ATMs from the dictionary"""
+        item_type: str = s.get("type", "")
+        if item_type == "atm":
+            filtered_atm_data.append(s)
+    atm_data = filtered_atm_data
```

Comment:
This (the filtering) can be included in `_get_atm_data()`. Then we can also cache the results so we don't read the file from disk each time a query is performed. I'd make `_get_atm_data` return the `filtered_atm_data` variable, and then add the following decorator:
```py
import cachetools

...

_TTL = 86400 # 1 day

@cachetools.cached(cachetools.TTLCache(1, _TTL))
def _get_atm_data():
    ...
```
This caches the ATM locations in memory for one day (86400 seconds).
We still need to do the distance calculations each time a query is made though.

Reply:
Great minds think alike.